miden-crypto/src/merkle/mmr/forest.rs

```diff
@@ -0,0 +1,245 @@
+use core::{
+    fmt::{Binary, Display},
+    ops::{BitAnd, BitOr, BitXor, BitXorAssign, ShlAssign},
+};
+
+use crate::Felt;
+
+/// A compact representation of trees (or peaks) in Merkle Mountain Range (MMR)
+///
+/// Each active bit of the stored number represents a disjoint tree with number of leaves
+/// equal to the bit position.
+///
+/// Examples:
+/// - Forest(0) is a forest with no trees.
+/// - Forest(0b01) is a forest with a single node (the smallest tree possible)
+/// - Forest(0b10) is a forest with a single binary tree with 2 leaves (3 modes)
+/// - Forest(0b11) is a forest with two trees: one with a single node, and one with 3 nodes
+/// - Forest(0b1010) is a forest with two trees: one with 8 leaves (15 nodes), one with 2 leaves (3
+///   nodes)
+#[derive(Debug, Copy, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Forest(usize);
+
+impl Forest {
+    /// Creates an empty forest (no trees)
+    pub const fn empty() -> Self {
+        Self(0)
+    }
+
+    /// Creates a forest with n leaves
+    pub const fn with_leaves(n: usize) -> Self {
+        Self(n)
+    }
```

Comment:
Could this just be `Forest::new`? Ideally we could rename `n` to `num_leaves` for clarity.